### PR TITLE
[feat] Add support for associating scheduler resources with an environment

### LINKED ISSUE #2991: Support injection of scheduler options from an environment
The concept that was discussed is the support of the following:

```
environs = [{
	'name': 'gnu',
	'sched_bind_options': ['--uenv-file={env.extras["uenv_file"]}',
						   '--uenv-mount={env.extras["uenv_mount"]']
}]
```

Using the above syntax, the `env` refers to the actual environment and use its extras.

### PR DESCRIPTION
**Pull request by ekouts**
This is a second attempt for this issue. ~I am missing the documentation + unittests but I would like first to agree on the api.~

It would be very similar to the resources of the test. And the configuration would look like this.

```python
site_configuration = {
    'systems': [
        {
            ...
            'partitions': [
                {
                    'resources': [
                        {
                            'name': 'switches',
                            'options': ['--switches={num_switches}']
                        },
                        {
                            'name': 'gres',
                            'options': ['--gres={gres}']
                        },
                        {
                            'name': 'uenv',
                            'options': [
                                '--uenv-file={file}',
                                '--uenv-mount={mount}',
                            ]
                        }
                    ]
                }
            ]
        }
    ],
    'environments': [
        {
            'name': 'myenv',
            'resources': {
                'uenv': {
                    'mount': 'one',
                    'file': 'two',
                }
            }
        }
    ]
}
```

That would translate to the following options in the job script:
```bash
#SBATCH --uenv-file=two
#SBATCH --uenv_mount=one
```

Closes #2991 .

### CODE REVIEW

File: docs/config_reference.rst [hunk 1]
```diff
@@ -766,6 +766,27 @@ ReFrame allows you to define custom scheduler resources for each partition that
     The backend assumes a ``qsub`` option, if the options passed in these attributes start with a ``-``.
 
 
+.. py:attribute:: systems.partitions.env_resources.name
+
+   :required: Yes
+
+  The name of this resources.
```
Comment (teojgo) [inline docs/config_reference.rst:773]:
```suggestion
  The name of the resources.
```

File: docs/config_reference.rst [hunk 2]
```diff
@@ -684,7 +684,7 @@ ReFrame can launch containerized applications, but you need to configure properl
 Custom Job Scheduler Resources
 ==============================
 
-ReFrame allows you to define custom scheduler resources for each partition that you can then transparently access through the :attr:`~reframe.core.pipeline.RegressionTest.extra_resources` attribute of a regression test.
+ReFrame allows you to define custom scheduler resources for each partition that you can then transparently access through the :attr:`~reframe.core.pipeline.RegressionTest.extra_resources` attribute of a regression test or the environment.
```
Comment (teojgo) [inline docs/config_reference.rst:687]:
```suggestion
ReFrame allows you to define custom scheduler resources for each partition/environment that can then be transparently accessed through the :attr:`~reframe.core.pipeline.RegressionTest.extra_resources` attribute of a regression test or the environment.
```

File: docs/config_reference.rst [hunk 3]
```diff
@@ -766,6 +766,27 @@ ReFrame allows you to define custom scheduler resources for each partition that
...
+   .. versionadded:: 4.6
+
+
+.. py:attribute:: systems.partitions.env_resources.options
+
+   :required: No
+   :default: ``[]``
+
+   A list of options to be passed to this partition’s job scheduler.
+   This is very similar to the :attr:`~config.systems.partitions.resources.options` parameter, but it is used to define resources that are specific to a programming environment.
```
Comment (teojgo) [inline docs/config_reference.rst:785]:
```suggestion
   This is very similar to the :attr:`~config.systems.partitions.resources.options` parameter, but it is used to define resources specific to a programming environment.
```

**Review by teojgo — CHANGES_REQUESTED** [review]

**Review by teojgo — APPROVED** [review]

File: reframe/core/environments.py [hunk 4]
```diff
@@ -326,3 +328,13 @@ def nvcc(self):
         :type: :class:`str`
         '''
         return self._nvcc
+
+    @property
+    def resources(self):
+        '''The resources associated with this environment.
```
Comment (vkarak) [inline reframe/core/environments.py:334]:
```suggestion
        '''The scheduler resources associated with this environment.
```

**Review by vkarak — CHANGES_REQUESTED** [review]
I think we should omit completely the `env_resources` and simply enable environments to use any of the scheduler resources. This would simplify the implementation and make also more sense at the partition level, as conceptually it doesn't make sense to separate the two types of resources.

File: docs/config_reference.rst [hunk 5]
```diff
@@ -689,7 +689,7 @@ ReFrame can launch containerized applications, but you need to configure properl
 Custom Job Scheduler Resources
 ==============================
 
-ReFrame allows you to define custom scheduler resources for each partition that you can then transparently access through the :attr:`~reframe.core.pipeline.RegressionTest.extra_resources` attribute of a regression test.
+ReFrame allows you to define custom scheduler resources for each partition/environment that can then be transparently accessed through the :attr:`~reframe.core.pipeline.RegressionTest.extra_resources` attribute of a regression test or the environment.
```
Comment (vkarak) [inline docs/config_reference.rst:692]:
```suggestion
ReFrame allows you to define custom scheduler resources for each partition that can then be transparently accessed through the :attr:`~reframe.core.pipeline.RegressionTest.extra_resources` attribute of a test or from an environment.
```

File: docs/config_reference.rst [hunk 6]
```diff
@@ -953,6 +953,16 @@ They are associated with `system partitions <#system-partition-configuration>`__
    It first looks for definitions for the current partition, then for the containing system and, finally, for global definitions (the ``*`` pseudo-system).
 
 
+.. py:attribute:: environments.resources
+
+   :required: No
+   :default: ``{}``
+
+   This is similar to a regression test's :attr:`~reframe.core.pipeline.RegressionTest.extra_resources`.
```
Comment (vkarak) [inline docs/config_reference.rst:961]:
```suggestion
   Scheduler resources associated with this environments.
   
   This is the equivalent of a test's :attr:`~reframe.core.pipeline.RegressionTest.extra_resources`.
```

File: unittests/resources/config/settings.py [hunk 7]
```diff
@@ -61,6 +61,13 @@ def hostname():
                                 '#DW jobdw capacity={capacity}',
                                 '#DW stage_in source={stagein_src}'
                             ]
+                        },
+                        {
+                            'name': 'uenv',
+                            'options': [
+                                '--mount={mount}',
+                                '--file={file}'
+                            ],
```
Comment (vkarak) [inline unittests/resources/config/settings.py:70]:
```suggestion
                            ]
```

**Review by vkarak — CHANGES_REQUESTED** [review]
Just some minor style changes.

**Review by vkarak — APPROVED** [review]

**Comment by ekouts** [comment]
Thanks @vkarak for finishing the PR!

**Pull request merged**